txstratum/api.py

```diff
@@ -112,21 +112,24 @@ async def submit_job(self, request: web.Request) -> web.Response:
             self.log.debug('non-standard-tx', data=data)
             return web.json_response({'error': 'non-standard-tx'}, status=400)
 
-        # Check for banned inputs.
-        for txin in tx.inputs:
-            if txin.tx_id in self.banned_tx_ids:
-                self.log.info('txin-banned', data=data)
+        if self.toiclient:
+            # Check for banned txs and addresses.
+            txs: Set[str] = set()
+            addrs: Set[str] = set()
+            for txin in tx.inputs:
+                txs.add(txin.tx_id.decode())
+
+            for txout in tx.outputs:
+                p2pkh = P2PKH.parse_script(txout.script)
+                if p2pkh is not None:
+                    self.log.info('p2pkh.address', address=p2pkh.address)
+                    addrs.add(p2pkh.address)
+
+            toiresp = await self.toiclient.check_blacklist(tx_ids=list(txs), addresses=list(addrs))
```

Comment:
If ToI service is down for some reason we will block all mining for transactions?

Reply:
For what I've seen it will return `blacklisted=False`, right? Anyway, I think we should have a short timeout (maybe 1s or less) otherwise we could increase the send transaction time if this service is fully loaded.

We should also have an alert if we have many timeouts in order to check why the service can't reply the requests in less than the timeout.

Reply:
I prefer having a retry mechanism (if the client using the tx-mining-service has some more room for waiting) and fallback to blocking if it fails to get a response from the toi-service.

Reply:
The opposite makes more sense to me because our capacity to _allow_ a transaction would depend on how resilient our services are. Which is how it naturally is. Otherwise in a chaotic high load situation, some transactions fails (because of overload) and some are allowed, but the ones that are allowed maybe should have been blocked, but weren't because we couldn't reach the toi-service.

What do you think @r4mmer?

Reply:
> A user that is blacklisted can try a DoS to bypass the blockage.

I think I agree with @jansegre in this situation, it makes sense to me given the discussion that we could block by default and it's extremely important that our alert system is used in order to understand if mining is being blocked because the service is taking too long to reply or if it's receiving many requests.

Reply:
Just a comment about the alert system that was mentioned. If we want to send alerts, all that's needed is to send messages in the right format to a SQS queue.

I can help with this if needed.

Reply:
I think Jan made good considerations about it. I feel we should have a way to enable/disabled a circuit breaker as needed. This might be a flag in a Redis, or just an administrative API to enable/disable the circuit breaker. For clarity, all transactions will be accepted by the TOI Filter when the circuit breaker is enabled.

Here are a few other ideas to discuss:

1) Should we keep track of the last N requests to the TOI Service and use it in case of a failure? We can use an internal redis for this so we can safely restart the tx-mining-service daemon.

2) Maybe we should add a metric of failure from the TOI Cliente. We can monitor this metric and alert accordingly.

Reply:
> I feel we should have a way to enable/disabled a circuit breaker as needed. This might be a flag in a Redis, or just an administrative API to enable/disable the circuit breaker.

It makes sense to have a way to enable/disable in runtime, but I think using an external service (database or otherwise) creates the same problem again (even if the service is much less likely to fail). I'd rather it be in-memory and use an internal API to change the state in memory, while the default state would be passed either from a command line argument, environment variable, configuration file, whatever fits best (if we have other parameters just adding a new one with the same mechanism).

> Here are a few other ideas to discuss:
> 
>     1. Should we keep track of the last N requests to the TOI Service and use it in case of a failure? We can use an internal redis for this so we can safely restart the tx-mining-service daemon.

I think a local LRU cache is good and simple enough. And that would be beneficial regardless of the circuit-break enabled state.

>     2. Maybe we should add a metric of failure from the TOI Cliente. We can monitor this metric and alert accordingly.

This could be very useful to guide us to change or reinforce our decision on the default circuit-break state. If the TOI service ends up being extremely stable and not fail even when the network is under a lot of stress, it would be really support not enabling the circuit-break, otherwise if it ends up being common to have random failures, specially if under stress before the other services, than the circuit-break would be much more justifiable IMO.

What do you think?